4단계 - 자동차 경주(우승자)

traumereiii opened the pull request:
안녕하세요 리뷰어님
오늘도 리뷰 잘 부탁 드립니다 :)

p.s 이전 PR에 남겨 놓은 코멘트에 대한 답변도 부탁 드릴게요!

DongSeong-Seo committed:
chore: 패키지명 변경

- step2 -> calculator
- step3 -> racinggame

DongSeong-Seo committed:
feat: #1085 PR 리뷰 반영

- RacingGame 클래스의 ui 참조 제거
- Car 클래스의 ui 참조 제거

DongSeong-Seo committed:
feat: ui클래스에 대한 인터페이스 제거 및 object 클래스로 변경

DongSeong-Seo committed:
feat: 자동차 게임 우승자 기능 추가

- Car 클래스에 이름 추가
- Car 클래스의 move 메서드 결과 쿨래스 추가(CarPosition)
- RacingGameResult 클래스 추가
  - 우승자
  - 단계별 결과

DongSeong-Seo committed:
feat: 자동차 게임 우승자 기능 추가

- Car 클래스 생성 시, 이름에 대한 Validation 추가

ohgillwhan commented on src/main/kotlin/study/racinggame/Main.kt:
```diff
@@ -0,0 +1,16 @@
+
+    println("실행 결과")
+
+    val racingGameResult = racingGame.start()
```
일반적으로 유추가 단순히 안되는 경우 타입을 명시해주는것도 좋습니다!

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/Car.kt:
```diff
@@ -0,0 +1,18 @@
+package study.racinggame.domain
+
+class Car(val name: String, private val engine: CarEngine) {
+    private var position: Int = 1
```
어떠한 자동차는 10부터 시작하는 조건이 있으면 어떻게 해야할까요?
생성자를 통해 값은 지정할 수 있게 하고 그 값이 비정상적인 것만 막아준다면 생성자는 풍부해도 괜찮다고 생각합니다.

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/Car.kt:
```diff
@@ -0,0 +1,18 @@
+
+    init {
+        if (name.isBlank() || name.length > 5) {
+            throw IllegalArgumentException()
```
`IllegalArgumentException` 💯 
조금 더 `custom`한 `Exception`을 만들면 테스트 코드에서 어떤 장점이 있을까요?

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/RacingGame.kt:
```diff
@@ -0,0 +1,34 @@
+    val racingGameStages: List<RacingGameStage>,
+    val winners: List<String>
+)
+data class RacingGameStage(
```
클래스 선언 사이에는 공백이 있으면 좋겠네요!

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/RacingGame.kt:
```diff
@@ -0,0 +1,34 @@
+data class RacingGameStage(
+    val stage: Int,
+    val positionOfCars: List<CarPosition>,
+
```
불필요한 공백은 제거해볼까요?

traumereiii replied:
ktlint에서 "Unexpected spacing before "(" (paren-spacing)" 메세지로 공백을 추가하지 말라는데 설정을 바꿔주어야 할까요?

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/RacingGame.kt:
```diff
@@ -0,0 +1,34 @@
+    val numberOfStages: Int,
+) {
+
+    fun start(): RacingGameResult {
+        val racingGameStages = (0 until numberOfStages)
+            .map { stage ->
+                val positionOfCars = cars.map { it.move() }
+                RacingGameStage(stage, positionOfCars)
+            }
+
+        val lastStageCarPositions = racingGameStages.last().positionOfCars
+        val maxPosition = lastStageCarPositions.map { it.position }
+            .maxBy { it }
+        val winners: List<String> = lastStageCarPositions
+            .filter { it.position == maxPosition }
+            .map { it.name }
+
+        return RacingGameResult(racingGameStages, winners)
+    }
```
15줄이 넘지 않았지만, 메서드가 조금 복잡해지기 시작하네요!
메서드 내에서 추상화를 하여 새로운 메서드를 만들면 어떤 장점이 있을까요? 관심사 분리가 가능하지 않을까요?

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/RacingGameConfiguration.kt:
```diff
@@ -0,0 +1,25 @@
+) {
+
+    fun build(): RacingGame {
+        val cars = nameOfCars.map { Car(it, RandomSeedEngine()) }
```
`RandomSeedEngine`은 싱글턴으로 가능하지 않을까요?

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/RacingGameConfiguration.kt:
```diff
@@ -0,0 +1,25 @@
+
+    companion object {
+        fun of(nameOfCars: List<String>, numberOfStages: Int): RacingGameConfiguration {
+            if (nameOfCars.size < 2) {
+                throw IllegalArgumentException()
+            }
+            if (numberOfStages < 1) {
+                throw IllegalArgumentException()
+            }
```
조금 더 `custom`한 `Exception`을 만들면 테스트 코드에서 어떤 장점이 있을까요?

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/RacingGameConfiguration.kt:
```diff
@@ -0,0 +1,25 @@
+                throw IllegalArgumentException()
+            }
+
+            return RacingGameConfiguration(nameOfCars, numberOfStages)
```
`return`을 할 때 한줄을 공백을 주셨네요! 다른곳도 통일해보면 어떨까요?
코틀린 공식 홈페이지에서 제공하지 않는 컨벤션은 스스로 만들어 보는것도 좋은것 같아요 👍

ohgillwhan commented on src/main/kotlin/study/racinggame/ui/CommandLineRacingGameConfigurator.kt:
```diff
@@ -0,0 +1,15 @@
+object CommandLineRacingGameConfigurator {
+    fun run(): RacingGameConfiguration {
+        println("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).")
+        val nameOfCars = readln().split(",")
```
","는 매직넘버와 같이 여러 의미로 해석될 가능성이 있어보여요. 상수로 뽑아보는건 어떨까요?
그리고 이 부분도 중요한 로직이라고 생각이 드는데, 테스트 할 수 없는게 조금 아쉽네요! 어떻게 하면 테스트 할 수 있을까요?

ohgillwhan reviewed (changes requested):
코드 잘 봤습니다! 💯 
리뷰 확인 부탁드릴게요! `step3`에서 궁금하셨던 점은 `DM`으로 답변 드렸습니다.
항상 궁금하신건 `DM`으로 주셔도 괜찮습니다. 파이팅!

DongSeong-Seo committed:
feat: PR 리뷰 반영 및 리팩토링

- CustomException 추가
  - InvalidParticipantNameException
  - CanNotNegativePositionException
  - InsufficientCarsException
  - InsufficientStagesException
- RacingGame.start 내부 메서드로 추상화
- RandomSeedEngine를 object 클래스로 변경

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/Car.kt:
```diff
@@ -0,0 +1,14 @@
+        return CarPosition(participant, position)
+    }
+}
+data class CarPosition(val participant: Participant, val position: Position)
```
클래스와 클래스 사이에 공백을 넣어볼까요?

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/Car.kt:
```diff
@@ -0,0 +1,14 @@
+) {
+
+    fun move(): CarPosition {
+        this.position = position.move(engine.run())
```
`Car`가 움직일지 말지 판단하고 `Position`을 변경해야 할것 같아요. 
`Position`이 움직임을 판단하고 움직이는건 조금 애매할 수 있을것 같아요.

일반적으로 우리가 생각하는 흐름을 코드 그대로 옮겨볼까요?

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/Participant.kt:
```diff
@@ -0,0 +1,11 @@
+package study.racinggame.domain
+
+data class Participant(val name: String) {
```
좋네요~ `value class`도 찾아볼까요?
https://velog.io/@dhwlddjgmanf/Kotlin-1.5%EC%97%90-%EC%B6%94%EA%B0%80%EB%90%9C-value-class%EC%97%90-%EB%8C%80%ED%95%B4-%EC%95%8C%EC%95%84%EB%B3%B4%EC%9E%90

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/Position.kt:
```diff
@@ -0,0 +1,15 @@
+            throw CanNotNegativePositionException(value)
+        }
+    }
+    fun move(value: Int = 1): Position {
```
함수와 함수 사이의 공백도 넣어볼까요?

ohgillwhan commented on src/main/kotlin/study/racinggame/domain/RacingGame.kt:
```diff
@@ -0,0 +1,38 @@
+        return RacingGameResult(racingGameStages, winners)
+    }
+
+    private fun chooseWinner(racingGameStages: List<RacingGameStage>): List<Participant> {
+        val lastStageCarPositions = racingGameStages.last().positionOfCars
+        val maxPosition = lastStageCarPositions.map { it.position }
+            .maxBy { it.value }
+        return lastStageCarPositions
+            .filter { it.position == maxPosition }
+            .map { it.participant }
+    }
```
좋네요 💯 
대신 `Winner`를 구하는 방식이 바뀐다면 `RacingGame`을 수정하는게 조금 아쉬울 수 있을것 같아요. `Winner`를 도출해보면 어떨까요?

ohgillwhan commented on src/main/kotlin/study/racinggame/ui/CommandLineRacingGameConfigurator.kt:
```diff
@@ -0,0 +1,14 @@
+object CommandLineRacingGameConfigurator {
+    private const val SEPARATOR: String = ","
+    fun run(): RacingGameConfiguration {
+        println("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).")
+        val nameOfCars = readln().split(SEPARATOR)
+        println("시도할 횟수는 몇 회인가요?")
+        val numberOfTries = readln()
+        return RacingGameConfiguration.of(nameOfCars, numberOfTries.toInt())
```
문단을 만들어서 공백을 조금 더 추가하면 좋을것 같아요.

ohgillwhan reviewed (approved):
코드 잘 작성 해주셨습니다.
다음 `step` 진행 부탁드립니다!
